# ANN Stock Trend Prediction With Large Amount of Predictor

## *This is a project 100% initiated by Max Hong Ka Ho and should solely be used for academic purpose only, any result in this passage should not constitute any investment advice*

### This project would use a large data set contributed by **Ehsan Hoseinzade, a Ph.D. student of computer science at Simon Fraser University** in which the scholar has used CNN to predict future prices. Here, I would like to borrow his data set for my own experiment as the owner of the data has offered a wide range of features for usage

### My Approach: we would like to use the neural network to approach this question. With the aid of the large data set, we are going to treat this as a binary classicfication problem where 1 indicates rises in tomorrow's price while 0 indicates a fall.

In [65]:
#Import library
import numpy as np
import pandas as pd
import tensorflow as tf

In [66]:
#read the data
data = pd.read_csv('Processed_S&P.csv')
data = data.set_index(data['Date']).iloc[:,1:]
print(data.head())

                  Close    Volume       mom  ...  wheat-F   XAG   XAU
Date                                         ...                     
2009-12-31  1115.099976       NaN       NaN  ...    -0.48  0.30  0.39
2010-01-04  1132.989990  0.921723  0.016043  ...     3.12  3.91  2.10
2010-01-05  1136.520020 -0.375903  0.003116  ...    -0.90  1.42 -0.12
2010-01-06  1137.140015  0.996234  0.000546  ...     2.62  2.25  1.77
2010-01-07  1141.689941  0.059932  0.004001  ...    -1.85  0.22 -0.58

[5 rows x 83 columns]


In [67]:
# check na values
data = data.fillna(0)
data.isna().sum()

Close             0
Volume            0
mom               0
mom1              0
mom2              0
                 ..
Dollar index-F    0
Dollar index      0
wheat-F           0
XAG               0
XAU               0
Length: 83, dtype: int64

In [68]:
#Create Return column
data['Return'] = (data['Close'] - data['Close'].shift(1)) / data['Close'].shift(1)
data['Return'].head()

Date
2009-12-31         NaN
2010-01-04    0.016043
2010-01-05    0.003116
2010-01-06    0.000546
2010-01-07    0.004001
Name: Return, dtype: float64

In [70]:
#Create Binary Column to denote the ups/downs of the S&P500
data['Trend'] = [1 if data['Return'][i] > 0 else 0 for i in range(0, len(data['Return']))]
data['Trend'] = data['Trend'].shift(-1)
data['Trend'].head(20)

Date
2009-12-31    1.0
2010-01-04    1.0
2010-01-05    1.0
2010-01-06    1.0
2010-01-07    1.0
2010-01-08    1.0
2010-01-11    0.0
2010-01-12    1.0
2010-01-13    1.0
2010-01-14    0.0
2010-01-15    1.0
2010-01-19    0.0
2010-01-20    0.0
2010-01-21    0.0
2010-01-22    1.0
2010-01-25    0.0
2010-01-26    1.0
2010-01-27    0.0
2010-01-28    0.0
2010-01-29    1.0
Name: Trend, dtype: float64

In [71]:
data.head()

Close    Volume       mom  ...   XAU    Return  Trend
Date                                         ...                       
2009-12-31  1115.099976  0.000000  0.000000  ...  0.39       NaN    1.0
2010-01-04  1132.989990  0.921723  0.016043  ...  2.10  0.016043    1.0
2010-01-05  1136.520020 -0.375903  0.003116  ... -0.12  0.003116    1.0
2010-01-06  1137.140015  0.996234  0.000546  ...  1.77  0.000546    1.0
2010-01-07  1141.689941  0.059932  0.004001  ... -0.58  0.004001    1.0

[5 rows x 85 columns]

In [72]:
# Create Independent and Dependent variables for our model
X = data.iloc[:, 14:-2].values[:-1]
y = data.iloc[:, -1].values[:-1]

In [73]:
print(X)

[[0.04 0.06 0.2 ... -0.48 0.3 0.39]
 [0.05 0.08 0.18 ... 3.12 3.91 2.1]
 [0.03 0.07 0.17 ... -0.9 1.42 -0.12]
 ...
 [1.03 1.21 1.34 ... 0.7 -0.71 -0.8]
 [1.04 1.22 1.35 ... -1.85 0.83 0.16]
 [1.04 1.24 1.37 ... 1.0 0.01 0.24]]


In [74]:
print(y)

[1. 1. 1. ... 1. 0. 0.]


In [75]:
# train test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [85]:
for i in range(len(X_train)):
  for j in range(len(X_train[0])):
    if type(X_train[i][j]) == str:
      X_train[i][j] = 0
    else:
      continue
for i in range(len(X_test)):
  for j in range(len(X_test[0])):
    if type(X_test[i][j]) == str:
      X_test[i][j] = 0
    else:
      continue

In [86]:
#feature scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [149]:
#Build the ANN
ann = tf.keras.models.Sequential()
ann.add(tf.keras.layers.Dense(units = 5, activation = 'relu'))
ann.add(tf.keras.layers.Dense(units = 5, activation = 'relu'))
ann.add(tf.keras.layers.Dense(units = 1, activation = 'sigmoid')) 

In [150]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [151]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)

Epoch 1/100
50/50 [==============================] - 1s 2ms/step - loss: 0.7783 - accuracy: 0.5013
Epoch 2/100
50/50 [==============================] - 0s 2ms/step - loss: 0.7177 - accuracy: 0.5259
Epoch 3/100
50/50 [==============================] - 0s 2ms/step - loss: 0.7008 - accuracy: 0.5309
Epoch 4/100
50/50 [==============================] - 0s 1ms/step - loss: 0.6938 - accuracy: 0.5473
Epoch 5/100
50/50 [==============================] - 0s 1ms/step - loss: 0.6888 - accuracy: 0.5536
Epoch 6/100
50/50 [==============================] - 0s 2ms/step - loss: 0.6863 - accuracy: 0.5580
Epoch 7/100
50/50 [==============================] - 0s 2ms/step - loss: 0.6826 - accuracy: 0.5574
Epoch 8/100
50/50 [==============================] - 0s 2ms/step - loss: 0.6808 - accuracy: 0.5624
Epoch 9/100
50/50 [==============================] - 0s 2ms/step - loss: 0.6789 - accuracy: 0.5694
Epoch 10/100
50/50 [==============================] - 0s 1ms/step - loss: 0.6781 - accuracy: 0.5738
Epoch 11/

In [152]:
# make predictions on test set
y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)
np.concatenate((y_pred.reshape(len(y_pred), 1), y_test.reshape(len(y_test), 1)), 1)

array([[0., 1.],
       [1., 1.],
       [1., 0.],
       [1., 1.],
       [1., 1.],
       [0., 0.],
       [1., 0.],
       [0., 1.],
       [1., 0.],
       [0., 0.],
       [0., 1.],
       [1., 1.],
       [1., 0.],
       [0., 1.],
       [1., 1.],
       [1., 1.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 1.],
       [0., 0.],
       [1., 1.],
       [1., 0.],
       [1., 0.],
       [1., 1.],
       [0., 1.],
       [1., 0.],
       [1., 1.],
       [1., 1.],
       [1., 0.],
       [1., 1.],
       [1., 1.],
       [1., 0.],
       [1., 1.],
       [0., 0.],
       [1., 1.],
       [0., 1.],
       [1., 1.],
       [1., 0.],
       [0., 0.],
       [1., 1.],
       [0., 0.],
       [1., 0.],
       [1., 0.],
       [1., 1.],
       [1., 0.],
       [0., 1.],
       [1., 1.],
       [1., 1.],
       [1., 1.],
       [1., 1.],
       [0., 1.],
       [1., 1.],
       [1., 1.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 1.],
       [1., 0.

In [153]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[ 58 131]
 [ 54 154]]


0.5340050377833753

# Conclusion

This experiment designed in above works poorly in test sample with maximum accuracy (after tried for lots of times) only 53.4% only, but 53% is a good score for financial predictions already and people may improve the model a bit to further enhance its accuracy